File: src/features/facts/sweep.ts
/**
 * Compaction of TTL ephemera, for maintenance rather than for reads.
 *
 * Cooldowns and other transient rows would otherwise grow the `flusk` log without
 * bound. Only rows explicitly marked transient AND already past their
 * `validUntil` may go: sweeping early would delete a fact that readers still
 * have to see, and sweeping a durable row would erase history no append can
 * rebuild. Nothing here changes what a read returns — expiry is decided from
 * the stored timestamp, so a row's removal is invisible by the time it happens.
 *
 * The rewrite replaces the file a writer would be appending to, so it takes
 * the same lock a transact does: a batch landing mid-sweep would be written
 * into the log the rename is about to discard.
 */
import { withLock } from "./lock.repository.js";
import { readLog, writeLog } from "./log.repository.js";
import { materialize } from "./materialize.js";
import { nsPath, storeDir } from "./paths.js";

/** Hard-deletes expired transient rows from `ns`; returns how many went. */
export async function sweepTransient(
	ns: string,
	at: number = Date.now(),
	dir: string = storeDir(),
): Promise<number> {
	const path = nsPath(dir, ns);
	return withLock(path, async () => {
		const records = await readLog(path);
		const doomed = new Set<string>();
		for (const row of materialize(records)) {
			if (!row.transient || row.fact.validUntil === null) continue;
			const until = Date.parse(row.fact.validUntil);
			if (Number.isFinite(until) && until <= at) doomed.add(row.fact.id);
		}
		if (doomed.size === 0) return 0;
		const kept = records.filter((r) => !doomed.has(r.k === "fact" ? r.fact.id : r.id));
		await writeLog(path, kept);
		return doomed.size;
	});
}

File: crates/flusk-core/src/store/sweep.rs
//! Compaction of TTL ephemera — src/features/facts/sweep.ts.
//!
//! Only rows explicitly marked transient AND already past their `validUntil`
//! may go: sweeping early would delete a fact readers still have to see, and
//! sweeping a durable row would erase history no append can rebuild. Nothing
//! here changes what a read returns — expiry is decided from the stored
//! timestamp, so a row's removal is invisible by the time it happens.
//!
//! The rewrite replaces the file a writer would be appending to, so it takes
//! the same lock a transact does: a batch landing mid-sweep would be written
//! into the log the rename is about to discard.

use super::jval::JVal;
use super::lock::with_lock;
use super::log::{read_log, write_log};
use super::materialize::materialize;
use super::time::parse_ms;
use std::collections::HashSet;
use std::path::Path;

/// Hard-deletes expired transient rows from the log at `path`; returns how
/// many went. The caller resolves the namespace to a path — the naming
/// convention (slug + hash) stays in TypeScript so there is one spelling.
pub fn sweep_transient(path: &Path, at: f64) -> Result<usize, String> {
    with_lock(path, || {
        let records = read_log(path)?;
        let mut doomed: HashSet<String> = HashSet::new();
        for row in materialize(&records) {
            if !row.transient {
                continue;
            }
            let Some(JVal::Str(until_iso)) = row.fact.get("validUntil") else {
                continue;
            };
            let until = parse_ms(until_iso);
            if until.is_finite() && until <= at {
                if let Some(id) = row.fact.get_str("id") {
                    doomed.insert(id.to_string());
                }
            }
        }
        if doomed.is_empty() {
            return Ok(0);
        }
        let kept: Vec<_> = records
            .into_iter()
            .filter(|r| !doomed.contains(r.subject_id()))
            .collect();
        write_log(path, &kept)?;
        Ok(doomed.len())
    })
}
